src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs

```diff
@@ -1341,5 +1346,51 @@ private Task<bool> InternalReceiveResponseHeadersAsync(WinHttpRequestState state
 
             return state.TcsReceiveResponseHeaders.Task;
         }
+
+        private static Guid LogRequest(HttpRequestMessage request)
+        {
+            if (_diagnosticSource.IsEnabled("System.Net.Http.Request"))
+            {
+                Guid loggingRequestId = Guid.NewGuid();
+                long timestamp = Stopwatch.GetTimestamp();
+
+                _diagnosticSource.Write(
+                    "System.Net.Http.Request",
+                    new
+                    {
+                        Request = request,
```

Comment:
This is one of those fundamental tradeoffs.   DiagnosticSource allows you to 'publish' lots of 'hooks' which show detailed state (note it should not be private class, but it is likely to be 'private' data).   Because we don't have a way of saying 'read only object' it also means that you CAN mutate things (although that is a abuse of DiagnosticSource to do so, the intention is that clients are effectively read-only).   

Thus DiagnosticSource creates a set of hooks that have 'scary power' like private reflection or the profiler API (both of which are not intended for program mutation).   It would be nice if it did not have this scary power to modify state, but there is no obvious way of doing that.    In a world of code access security Diagnostic source would have to be highly privileged.   

In short, we giving up on code access security (we trust all code in the process), and in that model this is OK.  

For what it is worth...